### PR TITLE
User login

### PR DESCRIPTION
**Pull request by pabloferro**
## Summary
  * User login.
  * Authenticated  views.

## Trello Card
https://trello.com/c/TNbzQu6p/3-crear-login-logout-de-usuarios

### CODE REVIEW

File: src/app/app.routes.js [hunk 1]
```diff
@@ -15,18 +15,53 @@ angular.module('pf-angular').config([
...
+          authenticate: false
         }
       });
-
     $locationProvider.html5Mode(true);
   }
 ]);
+
+
+angular.module('pf-angular').run([
```
Comment (sbalay) [inline src/app/app.routes.js]: Only declare one module per file.

Comment (sbalay) [inline src/app/app.routes.js]: To be clear. Declare this module on a new file.

File: src/app/app.routes.js [hunk 2]
```diff
@@ -15,18 +15,53 @@ angular.module('pf-angular').config([
...
+  '$rootScope', '$state', 'authService', 'Restangular', 'localStorageService',
+  function ($rootScope, $state, authService, Restangular, localStorageService) {
+    $rootScope.$on('$stateChangeStart', function (event, toState) {
+      if (toState.data.authenticate && !authService.isAuthenticated()) {
+        $state.transitionTo('login');
+        event.preventDefault();
+      }
+    });
+    Restangular.addFullRequestInterceptor(function (element, operation, route, url, headers) {
```
Comment (sbalay) [inline src/app/app.routes.js]: Querying the local storage may not be the most performant way of doing this. Can you think of any better option?

File: src/app/app.routes.js [hunk 3]
```diff
@@ -15,18 +15,53 @@ angular.module('pf-angular').config([
...
+        data: {
+          authenticate: true
         }
       })
       .state('signup', {
         url: '/signup',
-        controller: 'SignupController as signup',
         views: {
           main: {
+            controller: 'SignupController as signup',
```
Comment (sbalay) [inline src/app/app.routes.js]: Always end the controller aliases with ```Ctrl```. Just convention that makes reading the templates more clear.
```
controller: 'SignupController as signupCtrl',
```

File: src/app/app.routes.js [hunk 4]
```diff
@@ -15,18 +15,53 @@ angular.module('pf-angular').config([
...
+        },
+        data: {
+          authenticate: false
+        }
+      })
+      .state('login', {
+        url: '/login',
+        views: {
+          main: {
+            controller: 'LoginController as login',
```
Comment (sbalay) [inline src/app/app.routes.js]: end the alias with ```Ctrl```

File: src/app/components/login/loginController.js [hunk 5]
```diff
@@ -0,0 +1,15 @@
+angular.module('pf-angular').controller('LoginController', [
+  '$state', 'authService',
+  function ($state, authService) {
+    const that = this;
```
Comment (sbalay) [inline src/app/components/login/loginController.js]: This pattern gets deprecated with ecma6 arrow functions. Take a look at the [docs](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Functions/Arrow_functions#Lexical_this).

In particular, this is interesting about arrow functions:
> Arrow functions capture the this value of the enclosing context, so the following code works as expected.
```javascript
function Person(){
  this.age = 0;
  setInterval(() => {
    this.age++; // |this| properly refers to the person object
  }, 1000);
}
```

File: src/app/services/authService.js [hunk 6]
```diff
@@ -0,0 +1,24 @@
+angular.module('pf-angular').factory('authService', [
+  'Restangular', 'localStorageService',
+  function (Restangular, localStorageService) {
+    return {
+      login: (user) => {
+        return Restangular.all('login').customGET('', user).then(
+          function successCallback (response) {
```
Comment (sbalay) [inline src/app/services/authService.js]: Get used to use arrow functions instead of classical js functions.

File: src/app/app.routes.js (hunk 2 above)
Comment (pabloferro) [inline src/app/app.routes.js]: @sbalay can I have it in a var in the authService?

File: src/app/app.config.js [hunk 7]
```diff
@@ -1,13 +1,14 @@
 angular.module('pf-angular').config([
   'RestangularProvider', 'configuration', 'localStorageServiceProvider',
   function (RestangularProvider, configuration, localStorageServiceProvider) {
-
+    const appId = 'DfYLYMHIcC80ykwZFbQet4a3YqYkOD92hcXIMIfA';
```
Comment (sbalay) [inline src/app/app.config.js]: Put these keys in the environment config files.

File: src/app/app.run.js [hunk 8]
```diff
@@ -0,0 +1,16 @@
+angular.module('pf-angular').run([
+  '$rootScope', '$state', 'authService', 'Restangular',
+  function ($rootScope, $state, authService, Restangular) {
+    $rootScope.$on('$stateChangeStart', function (event, toState) {
+      if (toState.data.authenticate && !authService.isAuthenticated()) {
+        $state.transitionTo('login');
```
Comment (sbalay) [inline src/app/app.run.js]: ```$state.go('login')``` is preferred.
This is the definition of transitionTo
> Low-level method for transitioning to a new state. $state.go uses transitionTo internally. $state.go is recommended in most situations.

File: src/app/components/login/loginController.js [hunk 9]
```diff
@@ -0,0 +1,14 @@
+angular.module('pf-angular').controller('LoginController', [
+  '$state', 'authService',
+  function ($state, authService) {
+    this.submit = function (user) {
+      authService.login(user).then(
```
Comment (sbalay) [inline src/app/components/login/loginController.js]: We use ```.then``` for success callback and ```.catch``` for error calback. The code gets easier to read
```
authService.login(user).then(() => {
  $state.go('home');
}).catch((response) => {
  this.errorMessage = response.data.error;
})
```

File: src/index.jade [hunk 10]
```diff
@@ -11,6 +11,11 @@ html(lang='en', ng-app='pf-angular')
   body
     .container
       .center
+        .navbar(ng-controller='NavBarController as navbarCtrl')
```
Comment (sbalay) [inline src/index.jade]: This is ok, assuming that all the views in our app will have this navBar.
How would you do this if there are some states (for example, signUp) that do not show the navbar

Comment (sbalay) [inline src/index.jade]: Also, use the html5 element ```nav```

File: src/app/app.routes.js (hunk 2 above)
Comment (sbalay) [inline src/app/app.routes.js]: nice!

**Comment by sbalay** [comment]
Looks great! some new comments

File: src/index.jade (hunk 10 above)
Comment (pabloferro) [inline src/index.jade]: I could make it part of a abstract view that all views are children of (except login and signup).

**Comment by sbalay** [comment]
:+1:

**Pull request merged**